src/rabbit_dead_letter.erl

```diff
@@ -84,36 +84,85 @@ x_death_event_key(Info, Key, KeyType) ->
         {value, {Key, KeyType, Val}} -> Val
     end.
 
+maybe_append_to_event_group(Info, _Queue, _Reason, []) ->
+    [Info];
+maybe_append_to_event_group(Info, Queue, Reason, Acc) ->
+    case lists:any(queue_and_reason_matcher(Queue, Reason), Acc) of
+        true  -> Acc;
+        false -> [Info | Acc]
+    end.
+
+group_by_queue_and_reason(Tables) ->
+    Infos = [Info || {table, Info} <- Tables],
+    Grouped =
+        lists:foldl(fun (Info, Acc) ->
+            Q = x_death_event_key(Info, <<"queue">>, longstr),
+            R = x_death_event_key(Info, <<"reason">>, longstr),
+            Matcher = queue_and_reason_matcher(Q, R),
+            {Matches, _} = lists:partition(Matcher, Infos),
+            case Matches of
+                [X] ->
+                    maybe_append_to_event_group(
+                        ensure_xdeath_event_count(X), Q, R, Acc);
+                [X|_] = Xs when is_list(Xs) ->
+                    maybe_append_to_event_group(
+                        ensure_xdeath_event_count(X, length(Xs)), Q, R, Acc)
+            end
+        end, [], Infos),
+    [{table, Info} || Info <- Grouped].
```

Comment:
Currently it takes a proplist and returns a proplist. We could make it accept a pair and return a pair.